### PR TITLE
 transaction controller - use nonce-tracker

### PR DESCRIPTION
**Pull request by kumavis**

### CODE REVIEW

**Comment by frankiebee** [comment]
Here are some action steps me and Frankie came up with regarding this issue:

- [x] step 0:
remove retry limitation on pending txs, retry on block.

- [ ] step 1:
People have a big ugly queue of failed txs.
Those error’d txs should have the error icon & show failed.

- [ ] step 2:
calculate nonce based on local pending txs w/o error state.

- [ ] step 2.5:
Reset nonce switch to cancel pending: #1668

- [ ] step 3:
Clicking a pending tx goes to tx approval screen,
change “ACCEPT” to “SUBMIT” or “SEND”.

- [ ] step 4: allow editing nonce via an advanced thing on the tx approval screen.

**Comment by kumavis** [comment]
@frankiebee looking forward to merging this - can you update your check list
i dont think this requires ui to merge, that can be separate - better to get this in sooner

File: app/scripts/controllers/transactions.js [hunk 1]
```diff
@@ -173,43 +183,42 @@ module.exports = class TransactionController extends EventEmitter {
...
+      // approve
+      this.setTxStatusApproved(txId)
+      // get next nonce
+      const txMeta = this.getTx(txId)
+      const fromAddress = txMeta.txParams.from
+      nonceLock = await this.nonceTracker.getNonceLock(fromAddress)
+      txMeta.txParams.nonce = nonceLock.nextNonce
+      this.updateTx(txMeta)
+      // sign transaction
+      const rawTx = await denodeify(this.signTransaction.bind(this))(txId)
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: @frankiebee 
im not sure we're using this denodeify correctly (compared with local denodeify)

also - we should only have one denodeify

File: app/scripts/controllers/transactions.js [hunk 2]
```diff
@@ -237,16 +246,17 @@ module.exports = class TransactionController extends EventEmitter {
     })
   }
 
-  publishTransaction (txId, rawTx, cb = warn) {
+  publishTransaction (txId, rawTx) {
```
Comment (kumavis) [inline app/scripts/controllers/transactions.js]: prefer `async` than explicitly returning a promise

File: test/unit/tx-controller-test.js [hunk 3]
```diff
@@ -270,44 +270,37 @@ describe('Transaction Controller', function () {
...
+      .callsArgWithAsync(1, null, noop)
 
       const pubStub = sinon.stub(txController.txProviderUtils, 'publishTransaction')
-      .callsArgWith(1, null, originalValue)
-
-      txController.approveTransaction(txMeta.id, (err) => {
-        assert.ifError(err, 'should not error')
+      .callsArgWithAsync(1, null, originalValue)
 
+      return txController.approveTransaction(txMeta.id).then(() => {
```
Comment (kumavis) [inline test/unit/tx-controller-test.js]: missing a catch on the promise

**Comment by kumavis** [comment]
@frankiebee nice work frankie - please approve this pr and lets merge

**Review by frankiebee — APPROVED** [review]

**Pull request merged**